File: tests/test_utf8_truncate.c
/*
 * Test for UTF-8 aware truncation function
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../src/util/string_utils.h"

// Check if a string is valid UTF-8
int is_valid_utf8(const char *str) {
    const unsigned char *s = (const unsigned char *)str;
    while (*s) {
        if (*s < 0x80) {
            // ASCII (1 byte)
            s++;
        } else if ((*s & 0xE0) == 0xC0) {
            // 2-byte sequence: 110xxxxx 10xxxxxx
            if ((s[1] & 0xC0) != 0x80) return 0;
            s += 2;
        } else if ((*s & 0xF0) == 0xE0) {
            // 3-byte sequence: 1110xxxx 10xxxxxx 10xxxxxx
            if ((s[1] & 0xC0) != 0x80 || (s[2] & 0xC0) != 0x80) return 0;
            s += 3;
        } else if ((*s & 0xF8) == 0xF0) {
            // 4-byte sequence: 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
            if ((s[1] & 0xC0) != 0x80 || (s[2] & 0xC0) != 0x80 || (s[3] & 0xC0) != 0x80) return 0;
            s += 4;
        } else {
            return 0; // Invalid UTF-8
        }
    }
    return 1;
}

int main() {
    printf("Testing UTF-8 truncation function\n");
    printf("=================================\n\n");

    int passed = 0;
    int failed = 0;

    // Test 1: Simple ASCII
    {
        const char *input = "Hello World";
        char *result = truncate_utf8(input, 5);
        if (result && strcmp(result, "Hello") == 0 && is_valid_utf8(result)) {
            printf("✓ Test 1: ASCII truncation passed\n");
            passed++;
        } else {
            printf("✗ Test 1: ASCII truncation failed\n");
            failed++;
        }
        free(result);
    }

    // Test 2: UTF-8 arrows (→ is 3 bytes: E2 86 92)
    {
        const char *input = "A→B→C→D";  // A \xe2\x86\x92 B \xe2\x86\x92 C \xe2\x86\x92 D
        // Position 4 would split the first arrow (bytes: A=0, E2=1, 86=2, 92=3)
        char *result = truncate_utf8(input, 4);
        if (result && is_valid_utf8(result)) {
            printf("✓ Test 2: UTF-8 arrow truncation (4 bytes) passed: '%s'\n", result);
            passed++;
        } else {
            printf("✗ Test 2: UTF-8 arrow truncation (4 bytes) failed: '%s'\n", result);
            failed++;
        }
        free(result);
    }

    // Test 3: Truncate at exact UTF-8 boundary
    {
        // "A→B" = A + E2 86 92 + B = 5 bytes
        const char *input = "A→B";
        char *result = truncate_utf8(input, 1);
        if (result && is_valid_utf8(result)) {
            printf("✓ Test 3: Truncate to 1 byte passed: '%s'\n", result);
            passed++;
        } else {
            printf("✗ Test 3: Truncate to 1 byte failed: '%s'\n", result);
            failed++;
        }
        free(result);
    }

    // Test 4: String already within limit
    {
        const char *input = "Short";
        char *result = truncate_utf8(input, 100);
        if (result && strcmp(result, "Short") == 0) {
            printf("✓ Test 4: String within limit passed\n");
            passed++;
        } else {
            printf("✗ Test 4: String within limit failed\n");
            failed++;
        }
        free(result);
    }

    // Test 5: Empty string
    {
        const char *input = "";
        char *result = truncate_utf8(input, 10);
        if (result && strcmp(result, "") == 0) {
            printf("✓ Test 5: Empty string passed\n");
            passed++;
        } else {
            printf("✗ Test 5: Empty string failed\n");
            failed++;
        }
        free(result);
    }

    // Test 6: NULL input
    {
        char *result = truncate_utf8(NULL, 10);
        if (result == NULL) {
            printf("✓ Test 6: NULL input passed\n");
            passed++;
        } else {
            printf("✗ Test 6: NULL input failed\n");
            failed++;
            free(result);
        }
    }

    // Test 7: Multiple UTF-8 arrows in middle of truncation
    {
        // Create a long string with arrows that would be split at BASH_OUTPUT_MAX_SIZE
        char input[12235];
        memset(input, 'A', 12230);
        input[12230] = '\xe2';
        input[12231] = '\x86';
        input[12232] = '\x92';  // → arrow
        input[12233] = 'B';
        input[12234] = '\0';

        char *result = truncate_utf8(input, 12228);
        if (result && is_valid_utf8(result) && strlen(result) <= 12228) {
            printf("✓ Test 7: Large truncation at boundary passed (len=%zu)\n", strlen(result));
            passed++;
        } else {
            printf("✗ Test 7: Large truncation at boundary failed (len=%zu, valid=%d)\n",
                   result ? strlen(result) : 0, result ? is_valid_utf8(result) : 0);
            failed++;
        }
        free(result);
    }

    // Test 8: 4-byte UTF-8 (emoji)
    {
        // 😀 is 4 bytes: F0 9F 98 80
        const char *input = "A😀B😀C";
        // 6 bytes: A + emoji + B, truncate at 4 would split emoji
        char *result = truncate_utf8(input, 5);
        if (result && is_valid_utf8(result)) {
            printf("✓ Test 8: 4-byte UTF-8 truncation passed: '%s' (len=%zu)\n", result, strlen(result));
            passed++;
        } else {
            printf("✗ Test 8: 4-byte UTF-8 truncation failed: '%s'\n", result);
            failed++;
        }
        free(result);
    }

    printf("\n=================================\n");
    printf("Results: %d passed, %d failed\n", passed, failed);
    return failed > 0 ? 1 : 0;
}

File: zig/tests/test_utf8_truncate.zig
//! tests/test_utf8_truncate.zig — Zig port of tests/test_utf8_truncate.c
//!
//! Tests the UTF-8-aware truncation function `truncateUtf8` from
//! zig/util/string_utils.zig.
//!
//! The C version tested `truncate_utf8` from src/util/string_utils.c.
//! The Zig equivalent is `truncateUtf8` in zig/util/string_utils.zig.
//!
//! Ported test cases:
//!   1. ASCII-only string truncated exactly
//!   2. UTF-8 arrow (3-byte, U+2192 →) not split at a byte inside the sequence
//!   3. Truncation to 1 byte — lands before or at a multi-byte char boundary
//!   4. String already within limit returned whole
//!   5. Empty string returns empty string
//!   6. Large buffer with a multi-byte char near the limit — valid UTF-8 result
//!   7. 4-byte emoji (😀, U+1F600) not split

const std = @import("std");
const string_utils = @import("../util/string_utils.zig");

const truncateUtf8 = string_utils.truncateUtf8;

// ---------------------------------------------------------------------------
// helper: check that a byte slice is valid UTF-8
// ---------------------------------------------------------------------------

fn isValidUtf8(s: []const u8) bool {
    const view = std.unicode.Utf8View.init(s) catch return false;
    var it = view.iterator();
    while (it.nextCodepointSlice()) |_| {}
    return true;
}

// ---------------------------------------------------------------------------
// tests
// ---------------------------------------------------------------------------

test "truncateUtf8: ASCII-only string truncated at exact byte boundary" {
    const alloc = std.testing.allocator;

    const result = try truncateUtf8(alloc, "Hello World", 5);
    defer alloc.free(result);

    try std.testing.expectEqualStrings("Hello", result);
    try std.testing.expect(isValidUtf8(result));
}

test "truncateUtf8: string already within limit is returned whole" {
    const alloc = std.testing.allocator;

    const result = try truncateUtf8(alloc, "Short", 100);
    defer alloc.free(result);

    try std.testing.expectEqualStrings("Short", result);
}

test "truncateUtf8: empty string" {
    const alloc = std.testing.allocator;

    const result = try truncateUtf8(alloc, "", 10);
    defer alloc.free(result);

    try std.testing.expectEqualStrings("", result);
}

test "truncateUtf8: 3-byte UTF-8 arrow (→, U+2192) not split" {
    // "A→B→C→D" — the arrow → is E2 86 92 (3 bytes, 1 display column)
    // Truncating at 4 bytes would land inside the arrow sequence starting at
    // byte 1 (E2); we must back up to 1 byte ("A").
    const alloc = std.testing.allocator;

    const input = "A\xE2\x86\x92B\xE2\x86\x92C"; // A→B→C
    const result = try truncateUtf8(alloc, input, 4);
    defer alloc.free(result);

    try std.testing.expect(isValidUtf8(result));
    try std.testing.expect(result.len <= 4);
    // Must contain at least 'A'
    try std.testing.expect(result.len >= 1);
}

test "truncateUtf8: truncate to 1 byte stays valid" {
    // "A→B" = A + E2 86 92 + B = 5 bytes
    const alloc = std.testing.allocator;

    const input = "A\xE2\x86\x92B";
    const result = try truncateUtf8(alloc, input, 1);
    defer alloc.free(result);

    try std.testing.expect(isValidUtf8(result));
    try std.testing.expect(result.len == 1);
    try std.testing.expectEqualStrings("A", result);
}

test "truncateUtf8: large buffer with multi-byte char near limit" {
    // Build a 12234-byte string:
    //   12230 'A' bytes + E2 86 92 (→) + 'B' + NUL
    // Truncating at bash_output_max (12228) must not split the arrow.
    const alloc = std.testing.allocator;
    const bash_output_max: usize = 12_228;

    var input_buf = try alloc.alloc(u8, 12234);
    defer alloc.free(input_buf);
    @memset(input_buf[0..12230], 'A');
    input_buf[12230] = '\xE2';
    input_buf[12231] = '\x86';
    input_buf[12232] = '\x92'; // →
    input_buf[12233] = 'B';

    const result = try truncateUtf8(alloc, input_buf, bash_output_max);
    defer alloc.free(result);

    try std.testing.expect(isValidUtf8(result));
    try std.testing.expect(result.len <= bash_output_max);
    // At least some of the initial 'A' bytes must be present (the exact count
    // depends on where the UTF-8-safe cut lands, but must be >= 12226 at minimum)
    try std.testing.expect(result.len >= 12226);
}

test "truncateUtf8: 4-byte emoji (😀, U+1F600) not split" {
    // 😀 = F0 9F 98 80 (4 bytes)
    // "A😀B😀C"
    const alloc = std.testing.allocator;

    const input = "A\xF0\x9F\x98\x80B\xF0\x9F\x98\x80C"; // A😀B😀C
    // Truncate at 5 bytes: lands inside second emoji byte → back up to "A😀"
    const result = try truncateUtf8(alloc, input, 5);
    defer alloc.free(result);

    try std.testing.expect(isValidUtf8(result));
    try std.testing.expect(result.len <= 5);
    // At minimum "A" must be present
    try std.testing.expect(result.len >= 1);
}

test "truncateUtf8: 2-byte sequence (é, U+00E9) not split" {
    // "café" = c a f é = 3 ASCII + C3 A9 = 5 bytes total
    // Truncating at 4 bytes lands on 0xA9 (continuation) → backs up to 3 ("caf")
    const alloc = std.testing.allocator;

    const input = "caf\xC3\xA9"; // "café"
    const result = try truncateUtf8(alloc, input, 4);
    defer alloc.free(result);

    try std.testing.expectEqualStrings("caf", result);
    try std.testing.expect(isValidUtf8(result));
}

test "truncateUtf8: truncate at exact multi-byte boundary returns full char" {
    // "café" at 5 bytes (full string) — no truncation needed
    const alloc = std.testing.allocator;

    const input = "caf\xC3\xA9";
    const result = try truncateUtf8(alloc, input, 5);
    defer alloc.free(result);

    try std.testing.expectEqualStrings("caf\xC3\xA9", result);
    try std.testing.expect(isValidUtf8(result));
}
